src/plugin/csi/node.go

```diff
@@ -0,0 +1,239 @@
+)
+
+// NodeSvc implements this CSI NodeService GRPC service interface
+type NodeSvc struct {
+	id string
+}
+
+// NewNodeSvc constructs a NodeSvc instance
+func NewNodeSvc() *NodeSvc {
+	return &NodeSvc{
+		id: uuid.New(),
+	}
+}
+
+var (
+	// _pachClient is a Pachyderm client for gRPC endpoints to share. Don't
+	// access it directly, as it may still be 'nil' when the first few RPCs
+	// arrive
+	_pachClient    *client.APIClient
+	pachClientOnce sync.Once
+)
+
+// Initialize 'pachClient' for RPC endpoints
+// TODO(msteffen): handle auth-is-activated case
+func getPachClient() *client.APIClient {
+	pachClientOnce.Do(func() {
+		var err error
+		_pachClient, err = client.NewInCluster()
+		if err != nil {
+			panic(fmt.Sprintf("could not connect to Pachyderm: %v", err))
+		}
+	})
+	return _pachClient
+}
+
+func init() {
+	getPachClient() // start connecting to Pachd on startup
+}
+
+// NodeStageVolume implements the corresponding endpoint in the CSI service
+// definition. In general, it's used to partition and format the disk and mount
+// the disk on a node global directory. We use it to download Pachyderm data
+// into a node global directory
+func (s *NodeSvc) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
+	// Check arguments (including publish context parameters)
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+	if len(req.GetStagingTargetPath()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Target path missing in request")
+	}
+	if len(req.PublishContext[RepoParam]) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "\"repo\" missing from request's publish context")
+	}
+	repo := req.PublishContext[RepoParam]
+	if len(req.PublishContext[CommitParam]) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "\"commit\" missing from request's publish context")
+	}
+	commit := req.PublishContext[CommitParam]
+
+	// If non-default 'path'is set, make sure it exists in Pachyderm
+	c := getPachClient()
+	commitPath := "/"
+	if len(req.PublishContext[PathParam]) > 0 {
+		commitPath = req.PublishContext[PathParam]
+		if _, err := c.ListFile(repo, commit, commitPath); err != nil {
+			return nil, status.Error(codes.InvalidArgument,
+				fmt.Sprintf("could not stat Pachyderm data at %s@%s:%s: %v", repo, commit, commitPath, err))
+		}
+	}
+
+	path := pathlib.Join(provisionRoot + req.VolumeId)
+	// see if staging path exists
+	var fi os.FileInfo
+	for {
+		var err error
+		fi, err = os.Stat(path)
+		if err == nil {
+			// TODO(msteffen): handle case where data is partially-synced.
+			if err := os.RemoveAll(path); err != nil {
+				return nil, status.Error(codes.Internal,
+					fmt.Sprintf("could not remove existing dir at %q: %v", path, err.Error()))
+			}
+			continue // retry stat()
+		} else if !os.IsNotExist(err) {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
```

Comment:
I'm a little confused by this loop, maybe I'm just misreading, is there a way to get out of it without erroring? And what's the need to loop the `Stat` and the `RemoveAll` it seems like after the first time through it'll definitely be in the state we expect.